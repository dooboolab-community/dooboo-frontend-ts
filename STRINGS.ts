const STRINGS = {
  en: { // English
    BUTTON: 'Button',
    SIGNUP: 'Sign Up',
    LOGIN: 'Login',
    LOGOUT: 'Logout',
    GOOGLE_LOGIN: 'Login with Google',
    FACEBOOK_LOGIN: 'Login with Facebook',
    EMAIL: 'Email',
    EMAIL_HINT: 'Write email address.',
    PASSWORD: 'Password',
    PASSWORD_HINT: 'Write password.',
    COMPLETE: 'Done',
    GOTO_Temp: 'No Page',
    BACK: 'Back',
    NAVIGATE: 'Navigate',
  },
  ko: { // Korean
    BUTTON: '버튼',
    SIGNUP: '회원가입',
    LOGIN: '로그인',
    LOGOUT: '로그아웃',
    GOOGLE_LOGIN: '구글 로그인',
    FACEBOOK_LOGIN: '페이스북 로그인',
    EMAIL: '이메일',
    EMAIL_HINT: '이메일 주소를 입력해주세요.',
    PASSWORD: '비밀번호',
    PASSWORD_HINT: '비밀번호를 입력해주세요.',
    COMPLETE: '완료',
    GOTO_Temp: '없는 페이지',
    BACK: '뒤로',
    NAVIGATE: '이동하기',
  },
  // ja: { // Japanese
  // },
  // zh: { // Chinese
  // },
  // es: { // Spanish
  // },
  // fr: { // French
  // },
  // id: { // Indonesian
  // },
};

const DEFAULT_LANG = 'en';
let lang = DEFAULT_LANG;

if (navigator) {
  lang = navigator.language.substr(0, 2);
}

export const getString = (str: string) => {
  str = STRINGS[lang]
    ? STRINGS[lang][str]
    : STRINGS[DEFAULT_LANG][str];

  if (!str) {
    str = '...';
  }
  return str;
};

export default STRINGS;
