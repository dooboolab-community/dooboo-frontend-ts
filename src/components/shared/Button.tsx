import React, { Component } from 'react';
import styled from 'styled-components';
import { WhiteButton, TransparentButton } from '../ui/Buttons';

interface IProps {
  id?: string;
  white?: boolean;
  imgSrc?: any;
  text?: string;
  onClick?: () => void;
  isLoading?: boolean;
}

const Text = styled.span`
  font-size: 14px;
  color: rgb(128, 109, 216);
`;

const LogoImg = styled.img`
  position: absolute;
  left: 16px;
  height: 20px;
  width: 20px;
  object-fit: cover
`;

const Spinner = styled.div`
  border: 4px solid #f3f3f3; /* Light grey */
  border-top: 4px solid #3498db; /* Blue */
  border-radius: 50%;
  width: 24px;
  height: 24px;
  animation: spin 1s linear infinite;
  @keyframes spin {
    0% { transform: rotate(0deg); }
    100% { transform: rotate(360deg); }
  }
`;

function Button(props: IProps) {
  const { white, onClick, imgSrc, text } = props;
  if (props.white) {
    return (
      <WhiteButton
        style={{ height: '60px' }}
        onClick={onClick}
      >
        {
          props.isLoading
            ? <Spinner id='spinner'/>
            : <div>
              {
                props.imgSrc
                  ? <LogoImg
                    src={props.imgSrc}
                  />
                  : null
              }
              <Text>{props.text}</Text>
            </div>
        }
      </WhiteButton>
    );
  }
  return (
    <TransparentButton
      style={{ height: '60px' }}
      onClick={onClick}
    >
      {
        props.isLoading
          ? <Spinner id='spinner'/>
          : <div>
            {
              props.imgSrc
                ? <LogoImg
                  src={props.imgSrc}
                />
                : null
            }
            <Text style={{ color: 'white' }}>{props.text}</Text>
          </div>
      }
    </TransparentButton>
  );
}

export default Button;
