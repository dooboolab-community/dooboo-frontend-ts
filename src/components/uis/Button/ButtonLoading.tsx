import type {ReactElement} from 'react';
import styled from '@emotion/styled';

const Spinner = styled.div`
  align-self: center;
  border: 4px solid #f3f3f3; /* Light grey */
  border-top: 4px solid #3498db; /* Blue */
  border-radius: 50%;
  width: 24px;
  height: 24px;
  animation: spin 1s linear infinite;
  @keyframes spin {
    0% {
      transform: rotate(0deg);
    }
    100% {
      transform: rotate(360deg);
    }
  }
`;

const ButtonLoading = (): ReactElement => {
  return <Spinner id="spinner" />;
};

export default ButtonLoading;
