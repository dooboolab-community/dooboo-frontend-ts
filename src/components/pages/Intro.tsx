import {ReactElement, useState} from 'react';

import Button from '../uis/Button';
import {IC_GOOGLE_W} from '../../utils/Icons';
import {User} from '../../types';
import UserCard from '../uis/UserCard';
import {device} from '../../theme';
import styled from '@emotion/styled';
import {useNavigate} from 'react-router-dom';
import {useRecoilState} from 'recoil';
import {useThemeContext} from '../../providers/ThemeProvider';
import {userRecoilState} from '../../recoil/atoms';

const Container = styled.div`
  display: flex;
  height: 100vh;
  width: 100vw;
  align-self: stretch;
  overflow: scroll;
  background: ${(props) => props.theme.background};

  flex-direction: column;
  justify-content: flex-start;
  align-items: center;
  overflow: hidden;
`;

const ButtonWrapper = styled.div`
  position: absolute;
  flex-direction: column;

  @media ${device.mobileS} {
    bottom: 40px;
    width: 85vw;
    align-self: center;
  }

  @media ${device.tablet} {
    width: 50vw;
    right: 60px;
    align-self: center;
    top: 400px;
  }
`;

function Intro(): ReactElement {
  // eslint-disable-next-line
  let timer: any;
  const navigate = useNavigate();
  const [_, setUser] = useRecoilState(userRecoilState);
  const {changeThemeType} = useThemeContext();
  const [isLoggingIn, setIsLoggingIn] = useState(false);

  const onLogin = (): void => {
    setIsLoggingIn(true);

    setUser(null);

    timer = setTimeout(() => {
      const newUser: User = {
        displayName: 'dooboolab',
        age: 30,
        job: 'developer',
      };

      setUser(newUser);
      setIsLoggingIn(false);
      clearTimeout(timer);
    }, 1000);
  };

  return (
    <Container>
      <UserCard />
      <ButtonWrapper>
        <Button
          testID="SIGN_IN"
          imgSrc={IC_GOOGLE_W}
          isLoading={isLoggingIn}
          onClick={(): void => onLogin()}
          text={<fbt desc="sign in">Sign In</fbt>}
        />
        <Button
          onClick={(): void => navigate('/temp', {})}
          text={<fbt desc="navigate">Navigate</fbt>}
        />
        <Button
          testID="CHANGE_THEME"
          onClick={(): void => changeThemeType()}
          text={<fbt desc="change theme">Change theme</fbt>}
        />
      </ButtonWrapper>
    </Container>
  );
}

export default Intro;
