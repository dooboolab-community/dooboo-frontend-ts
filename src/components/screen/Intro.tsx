import React, { Component } from 'react';
import { Redirect } from 'react-router-dom';
import styled from 'styled-components';

import Button from '../shared/Button';

import { device } from '../../theme';
import { AppProvider as Provider, AppConsumer, AppContext } from '../../providers';

import { IC_FACEBOOK_W, IC_GOOGLE_W } from '../../utils/Icons';

import { IUser } from '../../types';

import { getString } from '../../../STRINGS';

const Container = styled.div`
  display: flex;
  height: 100vh;
  width: 100vw;
  align-self: stretch;
  overflow: scroll;
  background: ${(props) => props.theme.background};

  flex-direction: column;
  justify-content: flex-start;
  align-items: center;
  overflow: hidden;
`;

const ContentWrapper = styled.div`
  display: flex;
  flex-direction: column;
  @media ${device.mobileS} {
    max-width: 768px;
    height: 100vh;
    width: 100vw;
    justify-content: center;
    align-items: center;
  }

  @media ${device.tablet} {
    justify-content: flex-start;
    align-items: flex-start;
    margin-top: 400px;
  }
`;

const ButtonWrapper = styled.div`
  position: absolute;
  flex-direction: column;

  @media ${device.mobileS} {
    bottom: 40px;
    width: 85vw;
    align-self: center;
  }

  @media ${device.tablet} {
    width: 50vw;
    right: 60px;
    align-self: center;
    top: 400px;
  }
`;

const Text = styled.span`
  font-size: 18px;
  line-height: 1.5;
  font-family: sans-serif;
  color: #e3e3e3;
`;

interface IProps {
  history: any;
  store?: any;
}

interface IState {
  isLoggingIn: boolean;
}

function Intro(props: IProps) {
  let timer: any;
  const { state, dispatch } = React.useContext(AppContext);
  const [isLoggingIn, setIsLoggingIn] = React.useState(false);

  const onLogin = () => {
    dispatch({ type: 'reset-user' });
    setIsLoggingIn(true);
    timer = setTimeout(() => {
      const user: IUser = {
        displayName: 'dooboolab',
        age: 30,
        job: 'developer',
      };
      dispatch({ type: 'set-user', payload: user });
      setIsLoggingIn(false);
      clearTimeout(timer);
    }, 1000);
  };

  const navigate = () => {
    const location: object = {
      pathname: '/404',
      state: {},
    };
    props.history.push(location);
  };

  return (
    <Container>
      <ContentWrapper>
        <Text>{state.user.displayName}</Text>
        <Text>{state.user.age ? state.user.age : ''}</Text>
        <Text>{state.user.job}</Text>
      </ContentWrapper>
      <ButtonWrapper>
        <Button
          imgSrc={IC_GOOGLE_W}
          isLoading={isLoggingIn}
          onClick={() => onLogin()}
          // white={true}
          text={getString('LOGIN')}
        />
        <Button
          data-testid='btn2'
          onClick={() => navigate()}
          white={true}
          text={getString('NAVIGATE')}
        />
      </ButtonWrapper>
    </Container>
  );
}

export default Intro;
