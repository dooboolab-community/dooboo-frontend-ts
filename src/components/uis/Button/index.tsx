import {CSSProperties, ReactElement} from 'react';

import ButtonLoading from './ButtonLoading';
import {ButtonPrimary} from '../Styles';
import styled from '@emotion/styled';

interface Props {
  testID?: string;
  id?: string;
  style?: CSSProperties;
  imgSrc?: string;
  text?: string | ReactElement;
  onClick?: () => void;
  isLoading?: boolean;
}

const ButtonWrapper = styled.div`
  flex: 1;
  height: 100%;

  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
`;

const PrimaryText = styled.span`
  font-size: 14px;
  color: ${({theme}) => theme.btnPrimaryFont};
  margin: auto;
`;

const LogoImg = styled.img`
  position: absolute;
  left: 20px;
  height: 20px;
  width: 20px;
  object-fit: cover;
`;

function Button(props: Props): ReactElement {
  const {testID, onClick, imgSrc, text, style, isLoading} = props;

  return (
    <ButtonPrimary data-testid={testID} style={style} onClick={onClick}>
      {isLoading ? (
        <ButtonLoading />
      ) : (
        <ButtonWrapper>
          {imgSrc ? <LogoImg src={imgSrc} /> : null}
          <PrimaryText>{text}</PrimaryText>
        </ButtonWrapper>
      )}
    </ButtonPrimary>
  );
}

Button.defaultProps = {
  style: {display: 'flex', height: '60px'},
};

export default Button;
