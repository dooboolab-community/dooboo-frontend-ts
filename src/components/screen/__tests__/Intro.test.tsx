import * as renderer from 'react-test-renderer';

import {
  RenderResult,
  act,
  cleanup,
  fireEvent,
  render,
} from '@testing-library/react';
import { createTestElement, history } from '../../../../test/testUtils';

import Button from '../../shared/Button';
import Intro from '../Intro';
import React from 'react';
import { getString } from '../../../../STRINGS';

const props = {};
const component = createTestElement(<Intro {...props} />);

let container;

beforeEach(() => {
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  document.body.removeChild(container);
  container = null;
});

// test for the container page in dom
describe('[Intro] screen rendering test', () => {
  let json: renderer.ReactTestRendererJSON;

  it('should render outer component and snapshot matches', () => {
    json = renderer.create(component).toJSON();
    expect(json).toMatchSnapshot();
  });
});

describe('[Intro] Interaction', () => {
  let rendered: renderer.ReactTestRenderer;
  let root: renderer.ReactTestInstance;
  let renderResult: RenderResult;

  afterEach(cleanup);

  it('should simulate [onLogin] click with testing library', () => {
    jest.useFakeTimers();
    renderResult = render(component);
    fireEvent.click(renderResult.getByText(getString('LOGIN')));
    expect(setTimeout).toHaveBeenCalledTimes(1);

    act(() => {
      jest.runAllTimers();
    });

    expect(clearTimeout).toHaveBeenCalledTimes(1);
  });

  it('should simulate [navigate] when clicked', () => {
    jest.spyOn(history, 'push');
    rendered = renderer.create(component);
    root = rendered.root;

    const buttons = root.findAllByType(Button);
    buttons[1].props.onClick();
    expect(history.push).toHaveBeenCalledTimes(1);
  });

  it('should change theme when [change theme] has been clicked', () => {
    renderResult = render(component);
    const btnChangeTheme = renderResult.getByText(getString('CHANGE_THEME'));
    const clickResult1 = fireEvent.click(btnChangeTheme);
    expect(clickResult1).toBe(true);
    const clickResult2 = fireEvent.click(btnChangeTheme);
    expect(clickResult2).toBe(true);
  });
});
