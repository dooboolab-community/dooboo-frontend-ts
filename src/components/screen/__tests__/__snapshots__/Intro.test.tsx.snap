// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Intro] screen rendering test should render outer component and snapshot matches 1`] = `
<div
  className="sc-EHOje bnuiUQ"
>
  <div
    className="sc-bZQynM iahEzA"
  >
    <span
      className="sc-htoDjs hdqfnc"
    >
      
    </span>
    <span
      className="sc-htoDjs hdqfnc"
    >
      
    </span>
    <span
      className="sc-htoDjs hdqfnc"
    >
      
    </span>
  </div>
  <div
    className="sc-gzVnrw gqpNDZ"
  >
    <button
      className="sc-bdVaJa jAKSux"
      onClick={[Function]}
      style={
        Object {
          "height": "60px",
        }
      }
    >
      <div>
        <img
          className="sc-bxivhb gzWPeD"
          src="test-file-stub"
        />
        <span
          className="sc-htpNat gUfFlb"
          style={
            Object {
              "color": "white",
            }
          }
        >
          Login
        </span>
      </div>
    </button>
    <button
      className="sc-bwzfXH jGtnoc"
      onClick={[Function]}
      style={
        Object {
          "height": "60px",
        }
      }
    >
      <div>
        <span
          className="sc-htpNat gUfFlb"
        >
          Navigate
        </span>
      </div>
    </button>
  </div>
</div>
`;
